README

.. contents::

Introduction
============

This tool provides an easy way for users to retrieve their SLCS certificates from a SWITCH SLCS server.

Usage
-----

Usage::

  $ slick-init --help
  Usage: slick-init [options] [idp]
  
  Options:
    -h, --help            show this help message and exit
    -d DIR, --storedir=DIR
                          find IdP(s) whose name or unique ID contain a
                          specified string
    -f SEARCHSTRING, --find=SEARCHSTRING
                          find IdP(s) whose name or unique ID contain a
                          specified string
    -i IDP, --idp=IDP     unique ID of the IdP used to log in
    -k, --key             prompt for key-passphrase (use Shibboleth password
                          by default)
    -l, --list            list all available IdP(s)
    -s SLCS, --slcs=SLCS  location of SLCS server (if not specified, use
                          SLCS_SERVER system variable or settings from
                          [storedir]/slcs-client.properties
    -w, --write           write the arguments specified on the command line to
                          a config file
    -v, --verbose         print status messages to stdout
    --debug               print alot of messages to stdout


Config File
-----------

The contents of a simple config file::

  $ cat ~/slcs-client.properties
  [slcs]
  idp = VPAC
  url = https://slcs1.arcs.org.au/SLCS/login

Install
=======

Ubuntu
------
::

  apt-get install python-setuptools python-m2crypto

  easy_install --index-url http://code.arcs.org.au/pypi/ slick

Centos5
-------

::

  $ yum install python-setuptools swig openssl-devel gcc subversion

  $ sudo easy_install virtualenv
  $ virtualenv slick
  $ cd slick

Once we activate the virtual envionment the PATH will be changed so that 
files within slick/bin/ will take precidence.

::

  $ . ./bin/activate
  (slick)$ svn co http://svn.osafoundation.org/m2crypto/tags/0.19.1/ m2crypto
  (slick)$ cd m2crypto
  (slick)$ python setup.py build_ext -I/usr/include/openssl install
  (slick)$ easy_install slick
  (slick)$ deactivate

Once the virtulenv is deactivated you can still run the command directly

::
  ./bin/slick-init



CHANGES
Changelog
=========

1.3 - Unrelased
----------------
* added generated sphinx documentation
* asks for passphrase by default
* updated version of arcs.shibboleth.client that supports SWITCH Discovery Service.
* removed search and list options for idps because that can now be done interactivly


1.2 (2009-07-31)
----------------
* Key is now stored with the mode 0600
* Certificate is now stored with the mode 0644
* updated README


1.1 (2009-07-30)
----------------
* changed the cosole size detection tobe python 2.4 compatable.


1.0 (2009-07-09)
----------------

* Initial release
* Uses arcs.gsi to build certificates.
* Shibboleth support




slick/client.py
##############################################################################
#
# Copyright (c) 2009 Victorian Partnership for Advanced Computing Ltd and
# Contributors.
# All Rights Reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
##############################################################################

from optparse import OptionParser
from urllib2 import urlparse
import os, sys
from os import path
import logging
import struct, fcntl, termios
from cookielib import MozillaCookieJar

from arcs.shibboleth.client import Shibboleth, CredentialManager, Idp
from arcs.gsi.slcs import slcs_handler as slcs
from settings import Settings, settings_options

spUri = "https://slcs1.arcs.org.au/SLCS/login"


def terminal_dimensions():
    fd = os.open(os.ctermid(), os.O_RDONLY)
    if not os.isatty(fd):
        return (0,0)
    return struct.unpack('hh', fcntl.ioctl(fd, termios.TIOCGWINSZ, '1234'))


def print_list_wide(items):
    lmax = max([len(x) for x in items]) + 1
    width = terminal_dimensions()[1]
    if width:
        col = width/lmax
        i = 1
        for item in items:
            if not i%col:
                print item
            else:
                print item.ljust(lmax),
            i = i + 1
    else:
        for item in items:
            print item

usage = "usage: %prog [options] [idp]"
parser = OptionParser(usage)

settings_options(parser)
parser.add_option("-f", "--find", dest="idp_search",
                  help="find IdP(s) whose name or unique ID contain a \
                  specified string",
                  metavar="SEARCHSTRING")
parser.add_option("-k", "--key", action='store_true',
                  help="use Shibboleth password as key passphrase")
parser.add_option("-w", "--write",
                  action="store_true",
                  help="write the idp specified on the command line to \
                  a config file")
parser.add_option("-v", "--verbose",
                  action="count",
                  help="print status messages to stdout")
parser.add_option("-V", "--version", action='store_true',
                  help="print version number and exit")

# Set up a specific logger with our desired output level
log = logging.getLogger('slick-client')
log_handle = logging.StreamHandler()
DEBUG_FORMAT = "%(asctime)s - %(levelname)s - %(message)s"


def main():
    try:
        (options, args) = parser.parse_args()

        if options.version:
            from common import version
            print version
            return

        if not path.exists(options.store_dir):
            os.mkdir(options.store_dir)

        formatter = None
        log_level = logging.WARNING # default
        if options.verbose == 1:
            formatter = "%(message)s"
            log_level = logging.INFO
        elif options.verbose >= 2:
            formatter = DEBUG_FORMAT
            log_level = logging.DEBUG

        # Set up basic configuration, out to stderr with a reasonable default format.
        if formatter:
            logging.basicConfig(level=log_level, format=formatter)
        else:
            logging.basicConfig(level=log_level)

        settings = Settings(options)
        spUri = settings.slcs
        config_idp = settings.idp

        # Cert cert using specific IdP
        idp = " ".join(args) or config_idp
        idp = Idp(idp)
        slcs_login_url = spUri
        c = CredentialManager()
        cj = MozillaCookieJar()
        shibopener = Shibboleth(idp, c, cj)
        slcsresp = shibopener.openurl(slcs_login_url)

        log.info('Writing to files')
        cert = slcs(slcsresp)
        key_path = path.join(options.store_dir, 'userkey.pem')
        if not options.key:
            def callback(verify=False):
                from getpass import getpass
                while 1:
                    p1=getpass('Enter passphrase(or none for idp password):')
                    if not p1:
                        p1 = c.get_password()
                        return p1
                    p2=getpass('Verify passphrase:')
                    if p1==p2:
                        return p1
                    print "Password doesn't match"
        else:
            def callback(verify=False):
                return c.get_password()

        cert.get_key()._key.save_pem(key_path, callback=callback)
        os.chmod(key_path, 0600)
        cert_path = path.join(options.store_dir, 'usercert.pem')
        cert_file = open(path.join(options.store_dir, 'usercert.pem'), 'w')
        cert_file.write(repr(cert))
        cert_file.close()
        os.chmod(cert_path, 0644)

        if options.write:
            log.info('Writing a config')
            settings.save()

        print "\nexport X509_USER_CERT=%s\nexport X509_USER_KEY=%s" % \
                (cert_path, key_path)
        return
    except KeyboardInterrupt:
        print "\Cancelled"
        return

    if len(sys.argv) == 1:
        parser.print_help()

if __name__ == '__main__':
    main()

